Make UndefinedFunction.__new__ accept the full type() function signature

asmeurer opened the pull request:
This is required to make pickling work with cloudpickle (in Python 3). It also
makes it possible to create an undefined function directly with

``` py
class f(AppliedUndef, metaclass=UndefinedFunction):
    __module__ = None
```

(although this is not recommended)

I've also cleaned up `UndefinedFunction.__new__` a bit (using `super()` instead of
calling `BasicMeta` directly, setting `__module__` in the module dictionary).

This fixes #11692 

I'll add some tests in a bit, but I want to make sure none of the tests fail with this change first.

asmeurer committed:
Make UndefinedFunction.__new__ accept the full type() function signature

This is required to make pickling work with cloudpickle (in Python 3). It also
makes it possible to create an undefined function directly with

class f(AppliedUndef, metaclass=UndefinedFunction):
    __module__ = None

(although this is not recommended)

I've also cleaned up UndefinedFunction.__new__ a bit (using super() instead of
calling BasicMeta directly, setting __module__ in the module dictionary).

This fixes #11692

asmeurer committed:
Remove unused import

asmeurer committed:
Use better name for third argument to UndefinedFunction.__new__

It's not a module dictionary, it's the class's __dict__.

richardotis commented on sympy/core/function.py:
```diff
@@ -776,9 +775,11 @@ class UndefinedFunction(FunctionClass):
-        ret.__module__ = None
+    def __new__(mcl, name, bases=(AppliedUndef,), __dict__=None, **kwargs):
+        __dict__ = __dict__ or {}
+        __dict__.update(kwargs)
```
Is there a chance that this could mutate something passed by the caller? If so I think you could flip it around to be `kwargs.update(__dict__)` and pass `kwargs` instead.

asmeurer replied:
They should be mutually exclusive. `__dict__` gets passed in by the cloudpickle unpickler, or when creating an UndefinedFunction class directly. `kwargs` gets passed in by `Function` (this lets you set attributes of the undefined function by passing them as keyword arguments to `Function`, e.g., `Function('f', is_positive=True)`).

asmeurer commented:
Looks like this doesn't break anything.

asmeurer committed:
Test pickle protocol 4 in test_pickling.py

asmeurer committed:
Test cloudpickle in test_pickling.py (if it is installed)

asmeurer committed:
Space after comma

asmeurer committed:
Don't skip the entire test when pickling a class

There may be other protocols that can still work.

asmeurer committed:
Don't mess with warnings in the pickling tests

Based on the comments, this was needed for Python 2.6, which we don't support
any more. No warnings seem to be emitted in Python 2.7 or 3.5 after this
change.

asmeurer committed:
Test pickling undefined functions for the protocols that support it

That is, copy, deepcopy, and cloudpickle in Python 3.

It actually works with protocol 4 (and possibly 3), but not when the
definition is in a function as it is here.

asmeurer commented:
This is ready for full review (unless there are more test failures).

The whole test pickling file could use some cleaning up, but I've only done a minimal amount of cleaning here. I may at some point in the indeterminate future try to do a pickling push, to clean up this file and make more things picklable.

richardotis commented:
LGTM. (Should we start using the 'review' feature?)

asmeurer commented:
Sure, you can use it.
